sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -2267,6 +2267,156 @@ case class FormatNumber(x: Expression, d: Expression)
   override def prettyName: String = "format_number"
 }
 
+object ToNumber {
+  def transfer(input: UTF8String, pattern: UTF8String): UTF8String = {
+    val inputStr = input.toString
+    val patternStr = pattern.toString.toUpperCase(Locale.ROOT).replaceAll("FM", "")
+    val hasSign = inputStr.startsWith("-") ||
+      (patternStr.contains('S') && inputStr.contains('-')) ||
+      (patternStr.contains("MI") && inputStr.contains('-')) ||
+      (patternStr.contains("PR") && inputStr.startsWith("<"))
+    val (newInputStr, newPatternStr) = if (hasSign) {
+      (inputStr.replaceAll("-|<|>", ""), patternStr.replaceAll("S|(MI)|(PR)|(FM)", ""))
+    } else {
+      (inputStr, patternStr.replaceAll("FM", ""))
+    }
+    val inputChars = newInputStr.toCharArray()
+    val patternChars = newPatternStr.toIterator
+    val builder = new UTF8StringBuilder
+    var indexOfString = 0
+    var hasPoint = false
+    patternChars.foreach { c =>
+      if (newInputStr.length > indexOfString) {
+        val currentChar = inputChars(indexOfString)
+        c match {
+          case '9' | '0' if Character.isDigit(currentChar) =>
+            builder.append(newInputStr.substring(indexOfString, indexOfString + 1))
+            indexOfString += 1
+          case '9' | '0' | ' ' if currentChar == '.' =>
+            if (builder.build.numChars == 0) {
+              builder.append("0")
+            }
+          case '9' | '0' if currentChar == ',' =>
+          case '.' | 'D' if currentChar.equals('.') =>
+            hasPoint = true
+            builder.append(newInputStr.substring(indexOfString, indexOfString + 1))
+            indexOfString += 1
+          case ',' | 'G' if currentChar.equals(',') =>
+            indexOfString += 1
+          case ',' | 'G' if Character.isDigit(currentChar) =>
+          case 'L' =>
+            while (Character.isLetter(inputChars(indexOfString)) ||
+              inputChars(indexOfString) == '$') {
```

Comment:
What happens if `indexOfString >= inputChars.length`?

Reply:
`indexOfString` from 0 to `inputChars.length - 1`. There exists one check.